linesearch.go

```diff
@@ -10,132 +10,135 @@ import (
 	"github.com/gonum/floats"
 )
 
-// Linesearch is a linesearch-based optimization method.
-// It consists of a NextDirectioner, which specifies the next linesearch method,
-// and a LinesearchMethod which performs the linesearch in the direction specified
-// by the NextDirectioner.
-type Linesearch struct {
+// LinesearchMethod encapsulates the common functionality of gradient-based
```

Comment:
I liked your description in the comments
Here would be my suggested comment.
"LinesearchMethod represents an abstract optimization method in which a function is optimized through successive line search optimizations. It consists of a NextDirectioner, which specifies the search direction at each iteration, and a Linesearcher which performs a linesearch along the search direction."